File: exercism/python/roman-numerals/roman_numerals.py
arabic_to_roman = {
    1000: 'M',
    500: 'D',
    100: 'C',
    50: 'L',
    10: 'X',
    5: 'V',
    1: 'I',
}

def roman(number):
    if number == 0:
        return ''

    for ar, rom in arabic_to_roman.items():
        if number >= 900 and number < 1000:
            return 'CM' + roman(number-900)

        if number >= 400 and number < 500:
            return 'CD' + roman(number-400)

        if number >= 90 and number < 100:
            return 'XC' + roman(number-90)

        if number >= 40 and number < 50:
            return 'XL' + roman(number-40)

        if ar - number == 1:
            return 'I' + rom + roman(number-(ar-1))

        if number // ar > 0:
            return rom + roman(number-ar)

File: exercism/go/roman-numerals/roman_numerals.go
package romannumerals

import (
	"fmt"
	"sort"
)

type arabicToRoman struct {
	arabic int
	roman  string
}

var mappings = [13]arabicToRoman{
	arabicToRoman{1000, "M"},
	arabicToRoman{900, "CM"},
	arabicToRoman{500, "D"},
	arabicToRoman{400, "CD"},
	arabicToRoman{100, "C"},
	arabicToRoman{90, "XC"},
	arabicToRoman{50, "L"},
	arabicToRoman{40, "XL"},
	arabicToRoman{10, "X"},
	arabicToRoman{9, "IX"},
	arabicToRoman{5, "V"},
	arabicToRoman{4, "IV"},
	arabicToRoman{1, "I"},
}

// ToRomanNumeral converts an arabic number to a roman one
func ToRomanNumeral(arabic int) (string, error) {
	if arabic <= 0 || arabic > 3000 {
		err := fmt.Errorf("The arabic numeral %d cannot be converted to a roman numeral", arabic)
		return "", err
	}
	numberMappings := len(mappings)

	roman := ""
	for arabic > 0 {
		idx := sort.Search(numberMappings, func(i int) bool {
			return arabic >= mappings[i].arabic
		})

		if idx == numberMappings {
			return "", fmt.Errorf("There was an error processing the arabic digit")

		}

		roman += mappings[idx].roman
		arabic -= mappings[idx].arabic
	}

	return roman, nil
}
